File: frontend/struct/src/functions/sortings/quickSort.ts
import { SortStep } from "@/interfaces/interface";
import { partition } from "@/utils/sortingUtils";

export const quickSortStep = (
  arr: number[],
  step: SortStep & { stack?: Array<{ low: number; high: number }> },
  ascending: boolean = true,
): {
  sortedArray: number[];
  step: SortStep & { stack: Array<{ low: number; high: number }> };
} => {
  const sortedArray = [...arr];
  let { stack } = step;

  if (!stack) {
    stack = [{ low: 0, high: sortedArray.length - 1 }];
  }

  if (stack.length > 0) {
    const { low, high } = stack.pop()!;

    if (low < high) {
      const pi = partition(sortedArray, low, high, ascending);

      stack.push({ low, high: pi - 1 });
      stack.push({ low: pi + 1, high });
    }
  }

  return {
    sortedArray,
    step: { ...step, stack },
  };
};

File: src/quick.sort.cpp
#include <algorithm>

template <typename T> int partition(T *arr, int low, int high, bool ascending) {
  T   pivot = arr[high];
  int i = low - 1;

  for (int j = low; j < high; ++j) {
    if ((ascending && arr[j] < pivot) || (!ascending && arr[j] > pivot)) {
      ++i;
      std::swap(arr[i], arr[j]);
    }
  }
  std::swap(arr[i + 1], arr[high]);
  return i + 1;
}

template <typename T>
void quickSort(T *arr, int low, int high, bool ascending = true) {
  if (low < high) {
    int pi = partition(arr, low, high, ascending);

    quickSort(arr, low, pi - 1, ascending);
    quickSort(arr, pi + 1, high, ascending);
  }
}
